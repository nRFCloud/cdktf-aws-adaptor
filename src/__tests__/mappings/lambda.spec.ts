import { LambdaAlias } from "@cdktf/provider-aws/lib/lambda-alias/index.js";
import { LambdaFunction } from "@cdktf/provider-aws/lib/lambda-function/index.js";
import { LambdaLayerVersionPermission } from "@cdktf/provider-aws/lib/lambda-layer-version-permission/index.js";
import { LambdaLayerVersion } from "@cdktf/provider-aws/lib/lambda-layer-version/index.js";
import { LambdaPermission } from "@cdktf/provider-aws/lib/lambda-permission/index.js";
import {
  CfnAlias,
  CfnFunction,
  CfnLayerVersion,
  CfnLayerVersionPermission,
  CfnPermission,
} from "aws-cdk-lib/aws-lambda";
import { describe } from "vitest";
import { itShouldMapCfnElementToTerraformResource, synthesizeElementAndTestStability } from "../helpers.js";

describe("Lambda mappings", () => {
  itShouldMapCfnElementToTerraformResource(
    CfnPermission,
    {
      action: "lambda:InvokeFunction",
      functionName: "my-function",
      principal: "sns.amazonaws.com",
      sourceArn: "arn:aws:sns:us-east-1:123456789012:my-topic-1a2b3c4d",
      eventSourceToken: "token",
      functionUrlAuthType: "IAM",
      principalOrgId: "123456789012",
      sourceAccount: "123456789012",
    },
    LambdaPermission,
    {
      action: "lambda:InvokeFunction",
      functionName: "my-function",
      principal: "sns.amazonaws.com",
      sourceArn: "arn:aws:sns:us-east-1:123456789012:my-topic-1a2b3c4d",
      eventSourceToken: "token",
      functionUrlAuthType: "IAM",
      principalOrgId: "123456789012",
      sourceAccount: "123456789012",
    },
  );

  itShouldMapCfnElementToTerraformResource(
    CfnLayerVersion,
    {
      compatibleArchitectures: ["x86_64"],
      compatibleRuntimes: ["nodejs12.x"],
      content: {
        s3Bucket: "my-bucket",
        s3Key: "my-key",
        s3ObjectVersion: "my-version",
      },
      description: "my description",
      layerName: "my-layer",
      licenseInfo: "my-license",
    },
    LambdaLayerVersion,
    {
      s3Bucket: "my-bucket",
      s3Key: "my-key",
      s3ObjectVersion: "my-version",
      compatibleArchitectures: ["x86_64"],
      compatibleRuntimes: ["nodejs12.x"],
      description: "my description",
      layerName: "my-layer",
      licenseInfo: "my-license",
    },
  );

  itShouldMapCfnElementToTerraformResource(
    CfnLayerVersionPermission,
    {
      action: "lambda:GetLayerVersion",
      principal: "123456789012",
      organizationId: "o-123456",
      layerVersionArn: "arn:aws:lambda:us-east-1:123456789012:layer:my-layer:1",
    },
    LambdaLayerVersionPermission,
    {
      action: "lambda:GetLayerVersion",
      principal: "123456789012",
      organizationId: "o-123456",
      layerName: "arn:aws:lambda:us-east-1:123456789012:layer:my-layer:1",
      versionNumber:
        `$\{tonumber(element(split(\":\", \"arn:aws:lambda:us-east-1:123456789012:layer:my-layer:1\"), 7))}` as unknown as number,
      statementId: "teststackawstackresourceC278F851",
    },
  );

  it("should map CfnAlias to LambdaAlias", () => {
    synthesizeElementAndTestStability(
      CfnAlias,
      {
        description: "description",
        functionName: "functionName",
        functionVersion: "functionVersion",
        name: "name",
        routingConfig: {
          additionalVersionWeights: [
            {
              functionVersion: "1",
              functionWeight: 2,
            },
            {
              functionVersion: "2",
              functionWeight: 1,
            },
          ],
        },
      },
      LambdaAlias,
      {
        description: "description",
        functionName: "functionName",
        functionVersion: "functionVersion",
        name: "name",
        routingConfig: {
          additionalVersionWeights: {
            "1": 2,
            "2": 1,
          },
        },
      },
    );
  });

  it("should map CfnFunction to LambdaFunction", () => {
    synthesizeElementAndTestStability(
      CfnFunction,
      {
        architectures: ["x86_64"],
        code: {
          imageUri: "imageUri",
          s3Bucket: "s3Bucket",
          s3Key: "s3Key",
          s3ObjectVersion: "s3ObjectVersion",
          zipFile: "zipFile",
        },
        deadLetterConfig: {
          targetArn: "targetArn",
        },
        description: "description",
        environment: {
          variables: {
            key: "value",
          },
        },
        fileSystemConfigs: [
          {
            arn: "arn",
            localMountPath: "localMountPath",
          },
        ],
        functionName: "functionName",
        handler: "handler",
        imageConfig: {
          command: ["command"],
          entryPoint: ["entryPoint"],
          workingDirectory: "workingDirectory",
        },
        kmsKeyArn: "kmsKeyArn",
        layers: ["layers"],
        memorySize: 1,
        packageType: "packageType",
        reservedConcurrentExecutions: 1,
        role: "role",
        codeSigningConfigArn: "codeSigningConfigArn",
        ephemeralStorage: {
          size: 1,
        },
        runtime: "runtime",
        timeout: 1,
        tracingConfig: {
          mode: "Active",
        },
        vpcConfig: {
          securityGroupIds: ["securityGroupIds"],
          subnetIds: ["subnetIds"],
        },
        snapStart: {
          applyOn: "1",
        },
        tags: [{
          key: "key",
          value: "value",
        }],
      },
      LambdaFunction,
      {
        s3Bucket: "s3Bucket",
        memorySize: 1,
        reservedConcurrentExecutions: 1,
        runtime: "runtime",
        timeout: 1,
        functionName: "functionName",
        handler: "handler",
        role: "role",
        codeSigningConfigArn: "codeSigningConfigArn",
        imageUri: "imageUri",
        architectures: ["x86_64"],
        deadLetterConfig: {
          targetArn: "targetArn",
        },
        description: "description",
        environment: {
          variables: {
            key: "value",
          },
        },
        ephemeralStorage: {
          size: 1,
        },
        fileSystemConfig: {
          arn: "arn",
          localMountPath: "localMountPath",
        },
        imageConfig: {
          command: ["command"],
          entryPoint: ["entryPoint"],
          workingDirectory: "workingDirectory",
        },
        kmsKeyArn: "kmsKeyArn",
        layers: ["layers"],
        packageType: "packageType",
        vpcConfig: {
          securityGroupIds: ["securityGroupIds"],
          subnetIds: ["subnetIds"],
        },
        publish: true,
        tracingConfig: {
          mode: "Active",
        },
        snapStart: {
          applyOn: "1",
        },
        s3Key: "s3Key",
        s3ObjectVersion: "s3ObjectVersion",
        tags: {
          key: "value",
        },
      },
    );
  });
});
