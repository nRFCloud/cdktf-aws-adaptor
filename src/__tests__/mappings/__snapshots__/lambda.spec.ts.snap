// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Lambda mappings > Should map AWS::Lambda::LayerVersion 1`] = `
"{
  \\"provider\\": {
    \\"aws\\": [
      {
        \\"alias\\": \\"us_east_1\\",
        \\"region\\": \\"us-east-1\\"
      }
    ]
  },
  \\"resource\\": {
    \\"aws_lambda_layer_version\\": {
      \\"resource_22C949BF\\": {
        \\"compatible_architectures\\": [
          \\"x86_64\\"
        ],
        \\"compatible_runtimes\\": [
          \\"nodejs12.x\\"
        ],
        \\"description\\": \\"my description\\",
        \\"layer_name\\": \\"my-layer\\",
        \\"license_info\\": \\"my-license\\",
        \\"s3_bucket\\": \\"my-bucket\\",
        \\"s3_key\\": \\"my-key\\",
        \\"s3_object_version\\": \\"my-version\\",
        \\"skip_destroy\\": true
      }
    }
  },
  \\"terraform\\": {
    \\"backend\\": {
      \\"local\\": {
        \\"path\\": \\"/terraform.test-stack.tfstate\\"
      }
    },
    \\"required_providers\\": {
      \\"aws\\": {
        \\"source\\": \\"aws\\",
        \\"version\\": \\"5.37.0\\"
      }
    }
  },
  \\"variable\\": {
    \\"resource_refs_CDA17697\\": {
      \\"default\\": \\"\${aws_lambda_layer_version.resource_22C949BF.arn}\\",
      \\"type\\": \\"string\\"
    }
  }
}"
`;

exports[`Lambda mappings > Should map AWS::Lambda::LayerVersionPermission 1`] = `
"{
  \\"provider\\": {
    \\"aws\\": [
      {
        \\"alias\\": \\"us_east_1\\",
        \\"region\\": \\"us-east-1\\"
      }
    ]
  },
  \\"resource\\": {
    \\"aws_lambda_layer_version_permission\\": {
      \\"resource_22C949BF\\": {
        \\"action\\": \\"lambda:GetLayerVersion\\",
        \\"layer_name\\": \\"arn:aws:lambda:us-east-1:123456789012:layer:my-layer:1\\",
        \\"organization_id\\": \\"o-123456\\",
        \\"principal\\": \\"123456789012\\",
        \\"skip_destroy\\": true,
        \\"statement_id\\": \\"teststackawstackresourceC278F851\\",
        \\"version_number\\": \\"\${tonumber(element(split(\\\\\\":\\\\\\", \\\\\\"arn:aws:lambda:us-east-1:123456789012:layer:my-layer:1\\\\\\"), 7))}\\"
      }
    }
  },
  \\"terraform\\": {
    \\"backend\\": {
      \\"local\\": {
        \\"path\\": \\"/terraform.test-stack.tfstate\\"
      }
    },
    \\"required_providers\\": {
      \\"aws\\": {
        \\"source\\": \\"aws\\",
        \\"version\\": \\"5.37.0\\"
      }
    }
  },
  \\"variable\\": {
    \\"resource_refs_CDA17697\\": {
      \\"default\\": \\"\${aws_lambda_layer_version_permission.resource_22C949BF.id}\\",
      \\"type\\": \\"string\\"
    }
  }
}"
`;

exports[`Lambda mappings > Should map AWS::Lambda::Permission 1`] = `
"{
  \\"provider\\": {
    \\"aws\\": [
      {
        \\"alias\\": \\"us_east_1\\",
        \\"region\\": \\"us-east-1\\"
      }
    ]
  },
  \\"resource\\": {
    \\"aws_lambda_permission\\": {
      \\"resource_22C949BF\\": {
        \\"action\\": \\"lambda:InvokeFunction\\",
        \\"event_source_token\\": \\"token\\",
        \\"function_name\\": \\"my-function\\",
        \\"function_url_auth_type\\": \\"IAM\\",
        \\"principal\\": \\"sns.amazonaws.com\\",
        \\"principal_org_id\\": \\"123456789012\\",
        \\"source_account\\": \\"123456789012\\",
        \\"source_arn\\": \\"arn:aws:sns:us-east-1:123456789012:my-topic-1a2b3c4d\\"
      }
    }
  },
  \\"terraform\\": {
    \\"backend\\": {
      \\"local\\": {
        \\"path\\": \\"/terraform.test-stack.tfstate\\"
      }
    },
    \\"required_providers\\": {
      \\"aws\\": {
        \\"source\\": \\"aws\\",
        \\"version\\": \\"5.37.0\\"
      }
    }
  },
  \\"variable\\": {
    \\"resource_refs_CDA17697\\": {
      \\"default\\": \\"\${aws_lambda_permission.resource_22C949BF.id}\\",
      \\"type\\": \\"string\\"
    }
  }
}"
`;

exports[`Lambda mappings > should map CfnAlias to LambdaAlias 1`] = `
"{
  \\"provider\\": {
    \\"aws\\": [
      {
        \\"alias\\": \\"us_east_1\\",
        \\"region\\": \\"us-east-1\\"
      }
    ]
  },
  \\"resource\\": {
    \\"aws_lambda_alias\\": {
      \\"resource_22C949BF\\": {
        \\"description\\": \\"description\\",
        \\"function_name\\": \\"functionName\\",
        \\"function_version\\": \\"functionVersion\\",
        \\"name\\": \\"name\\",
        \\"routing_config\\": {
          \\"additional_version_weights\\": {
            \\"1\\": 2,
            \\"2\\": 1
          }
        }
      }
    }
  },
  \\"terraform\\": {
    \\"backend\\": {
      \\"local\\": {
        \\"path\\": \\"/terraform.test-stack.tfstate\\"
      }
    },
    \\"required_providers\\": {
      \\"aws\\": {
        \\"source\\": \\"aws\\",
        \\"version\\": \\"5.23.1\\"
      }
    }
  },
  \\"variable\\": {
    \\"resource_refs_CDA17697\\": {
      \\"default\\": \\"\${aws_lambda_alias.resource_22C949BF.id}\\",
      \\"type\\": \\"string\\"
    }
  }
}"
`;

exports[`Lambda mappings > should map CfnFunction to LambdaFunction 1`] = `
"{
  \\"provider\\": {
    \\"aws\\": [
      {
        \\"alias\\": \\"us_east_1\\",
        \\"region\\": \\"us-east-1\\"
      }
    ]
  },
  \\"resource\\": {
    \\"aws_lambda_function\\": {
      \\"resource_22C949BF\\": {
        \\"architectures\\": [
          \\"x86_64\\"
        ],
        \\"code_signing_config_arn\\": \\"codeSigningConfigArn\\",
        \\"dead_letter_config\\": {
          \\"target_arn\\": \\"targetArn\\"
        },
        \\"description\\": \\"description\\",
        \\"environment\\": {
          \\"variables\\": {
            \\"key\\": \\"value\\"
          }
        },
        \\"ephemeral_storage\\": {
          \\"size\\": 1
        },
        \\"file_system_config\\": {
          \\"arn\\": \\"arn\\",
          \\"local_mount_path\\": \\"localMountPath\\"
        },
        \\"function_name\\": \\"functionName\\",
        \\"handler\\": \\"handler\\",
        \\"image_config\\": {
          \\"command\\": [
            \\"command\\"
          ],
          \\"entry_point\\": [
            \\"entryPoint\\"
          ],
          \\"working_directory\\": \\"workingDirectory\\"
        },
        \\"image_uri\\": \\"imageUri\\",
        \\"kms_key_arn\\": \\"kmsKeyArn\\",
        \\"layers\\": [
          \\"layers\\"
        ],
        \\"memory_size\\": 1,
        \\"package_type\\": \\"packageType\\",
        \\"publish\\": true,
        \\"reserved_concurrent_executions\\": 1,
        \\"role\\": \\"role\\",
        \\"runtime\\": \\"runtime\\",
        \\"s3_bucket\\": \\"s3Bucket\\",
        \\"s3_key\\": \\"s3Key\\",
        \\"s3_object_version\\": \\"s3ObjectVersion\\",
        \\"snap_start\\": {
          \\"apply_on\\": \\"1\\"
        },
        \\"tags\\": {
          \\"key\\": \\"value\\"
        },
        \\"timeout\\": 1,
        \\"tracing_config\\": {
          \\"mode\\": \\"Active\\"
        },
        \\"vpc_config\\": {
          \\"security_group_ids\\": [
            \\"securityGroupIds\\"
          ],
          \\"subnet_ids\\": [
            \\"subnetIds\\"
          ]
        }
      }
    }
  },
  \\"terraform\\": {
    \\"backend\\": {
      \\"local\\": {
        \\"path\\": \\"/terraform.test-stack.tfstate\\"
      }
    },
    \\"required_providers\\": {
      \\"aws\\": {
        \\"source\\": \\"aws\\",
        \\"version\\": \\"5.37.0\\"
      }
    }
  },
  \\"variable\\": {
    \\"resource_refs_CDA17697\\": {
      \\"default\\": \\"\${join(\\\\\\"\\\\\\", [aws_lambda_function.resource_22C949BF.arn, \\\\\\",\\\\\\", \\\\\\"[object Object]\\\\\\", \\\\\\",\\\\\\", aws_lambda_function.resource_22C949BF.snap_start[0].apply_on, \\\\\\",\\\\\\", aws_lambda_function.resource_22C949BF.snap_start[0].optimization_status])}\\",
      \\"type\\": \\"string\\"
    }
  }
}"
`;
